File: RequestToMonday_grapQL-3.py
import requests
import json
import pandas as pd


apiKey = ""

apiUrl = "https://api.monday.com/v2"
headers = {"Authorization": apiKey}

#GraphQL query to retrieve data from Monday.com
query2 = 'query {  boards(limit:1,ids:[3894196945]) {  name  groups(ids:["topics"]){ title items{  name   column_values{  title  text}}}}}'

data = {'query': query2}

json_data = json.loads(requests.post(
    url=apiUrl, json=data, headers=headers).text)


number_of_columns = len(json_data['data']['boards'][0]
                        ['groups'][0]['items'][0]['column_values'])

print("\n\nNumber of columns: ", number_of_columns)


titles = []
for i in range(number_of_columns-1):  # grab column titles from Monday.com response
    titles.append(json_data['data']['boards'][0]['groups']
                  [0]['items'][0]['column_values'][i]['title'])

# the response from Monday.com misses the first column, so we have to add it manually
headers_excel = ['Name']
# merge the column titles with the first header "Name"
headers_excel = headers_excel + titles

print("Headers_excel: ", headers_excel)  # print to see the result

num_items = len(json_data['data']['boards'][0]
                ['groups'][0]['items'][0]['column_values'][0]['text'])

print("\n\nNumber of items: ", num_items)

exit()

data = [[item['name']] + [c_v['text']
                          for c_v in item['column_values']]
        for item in json_data['data']['boards'][0]['groups'][0]['items'][0]['column_values']]


df = pd.DataFrame(data)

num_rows = df.shape[0]  # count number of rows
print("\nNumber of rows:", num_rows)
print("\n\nData: ", data)


print("\n\ndf: ", df)


file = open("data.csv", "w")
file.truncate(0)  # clean the all columns in the csv file
file.close()
df.to_csv(r'data.csv', index=None, sep=";")


print("\ndf: ", df)
df.to_excel("data2.xlsx", index=None, header=headers_excel)

File: RequestToMonday_grapQL-3.js
//main function
// the API version written in this code is depricated, pls consult the new version on the Monday.com wibsite
function requestMonday() {

  let base_url= "https://api.monday.com/v2";
  let apiKey = "";

  // query just to get board name, groups ids and names
  let query = `query {
    boards(limit:1,ids:[120818187]) { 
        name
            groups{
              id
              title
                  }
          }
        }`;

  let headers = {
      "Authorization": apiKey,
      "Content-Type": "application/json"
    }

  let options = {
      headers: headers,
      method: "POST",
      payload: JSON.stringify({query})
    }

  let response = UrlFetchApp.fetch(base_url,options).getContentText();
  let result = JSON.parse(response);

  //console.log("Group to be written: ",result.data.boards[0].groups[0].title) 

  let groups = result.data.boards[0].groups
  console.log(groups.title);

  //clear spreadsheet
  console.log("-= Clearing spreadsheet =-");
  clearSpreadsheet();

  //write headers
  console.log("-= Writing headers =-");
  writeHeaders();

  

  //write data
  for(let i = 0; i < groups.length;i++){
      writeSpreadsheet(groups[i].id);
       
  }
     
}


//to avoid too much data in one query, we query Monday.com API by group, and write in the googlesheet by group as well
function writeSpreadsheet(group){

  let base_url= "https://api.monday.com/v2";
  let apiKey = "";

  let page = 1
  let number_of_items = 20
  let groups = []
  let items = []
  let column_values = []

  while (number_of_items <= 20 && number_of_items > 0){
    let query = `query {
    boards(limit:1,ids:[120818187]) {
            groups(ids:[`+group+`]){
              title
              items (limit: 20, page: `+page+` ){
                  name
                  column_values{
                    text
                  }
          }
            }

      }
  }`;

    let headers = {
      "Authorization": apiKey,
      "Content-Type": "application/json"
    }

    let options = {
      headers: headers,
      method: "POST",
      payload: JSON.stringify({query})
    }

    let response = UrlFetchApp.fetch(base_url,options).getContentText();
    let result = JSON.parse(response);

    groups = result.data.boards[0].groups
    let items = groups[0].items  

    for (let j = 0; j < items.length; j++){    
      
      let new_row = []
      column_values = items[j].column_values
      new_row.push(groups[0].title)
      new_row.push(items[j].name)

      for (let k = 0; k < column_values.length; k++){  
            new_row.push(column_values[k].text)
      }

      SpreadsheetApp.getActiveSpreadsheet().getSheetByName("Monday export").appendRow(new_row);
      }
    
    number_of_items = items.length;
    console.log("Number of items: ", number_of_items);
    console.log("Page: ", page," for ",group);
    page++;
  }
}

function clearSpreadsheet(){
  //clear spreadsheet
  SpreadsheetApp.getActiveSpreadsheet().getSheetByName("Monday export").getRange("A1:FI10000").clearContent();
  console.log("-= spreadsheet cleaned =-");
}














